### PR TITLE
[FIX] website: ensure unicity of find

### PR DESCRIPTION
**Pull request by diagnoza**
Description of the issue/feature this PR addresses:
https://github.com/odoo/upgrade-specific/pull/1594

Current behavior before PR:
`find` treated as a single record, leading to `ValueError` at https://github.com/odoo/odoo/blob/14.0/addons/website/models/ir_module_module.py#L158 if it happens to contain more than one record.

Desired behavior after PR is merged:
<strike>Loop over `find` as it may contain more than one record.</strike>
Strengthen the search to ensure that find always contain one record.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by diagnoza** [comment]
I see your point. But if such situation should not occur under **any** circumstances, then why not have a unique constraint on `(key, website_id)`?


> Can't we delete the duplicates in a pre-migrate script instead?

That was my initial idea in https://github.com/odoo/upgrade-specific/pull/1594 - set `theme_template_id` on the duplicate(s) to `NULL`. I would have to write such script for each and every db that fails this way.

Another way would be to search for such duplicates in a `0.0.0` script (executed for every db, regardless of target version) and delete/nullify. Or a `website/15.0` `pre-` script, since apparently it happens only when upgrading to `15.0` (although on the other hand nothing proves that it cannot happen when migrating to `14.0`...).

**Comment by rdeodoo** [comment]
> then why not have a unique constraint on (key, website_id)?

There probably should have been, we didn't think/thought it was needed at time because no flow should lead to that except if the user plays in backend. But seeing your PR, we were thinking to add that in master.

> Another way would be to search for such duplicates in a 0.0.0 script (executed for every db, regardless of target version) and delete/nullify

AFAIC, that would be the best idea.

I'll let you both decide what you want to do here, but if you were to merge this PR, please add a comment on the line, something explaining this line is only for migration purpose[1] and will be replaced by a SQL constraint in saas-15.1.

-----

And about nullify or delete, I don't know what's best, we would have to check the DB to see if:
- both are siblings in the tree views or
- one is wrongly placed

in the first case, we can keep whichever one and delete/nullify other copies (unless there is children bellow one but not the others, where we should preserve the correct one (the one with children))
in the last case, only one of the copy would actually be correct and nullifying/removing the correct one would probably lead to weird behavior (eg theme upgrade not doing anything).

Case 1
```
        A
        |
        B'
     /      \
    C        C'
    |        |
    D        D
   / \       |
  E  E'      E'
```
Case 2
```
        A
        |
        B'
     /      \
    C        C'
    |        |
    D        D
    |       / \
    E      E'  E'
    |      |
    F      F'
```

For case 1, removing the specific view (E') inside the generic tree view (right side) would be incorrect, as it would leave the incorrect view in the DB, so the specific tree view would stop on D level.

For case 2, it is not problematic unless there is children bellow (F layer), where we should preserve the correct one (the E' with F' as child).

You get the idea.. Not sure you want to spend time on this tho, so maybe just nullify all duplicate copy and hope for the best.

Let me know if I am missing something or saying something wrong, I am trying to figure out what are the effect of those duplicate copies (besides the migration failure).

[1] Indeed, it is not fixing anything but just allow the code to get through (in migration in the current case), see my examples above, database would still be broken (if Case 1 and/or Case 2 is true).

**Comment by diagnoza** [comment]
I managed to reproduce the issue as follows:

1. Duplicate an attachment
2. Modify `ir.attachment` form view to display `theme_template_id`
3. Set `theme_template_id` to point to the same record on both original and duplicated attachment

After discussing it with @KangOl, I'd like this pr merged. I added a comment before the loop as requested. 
Also please keep in mind, that the unique constraint would have to be on `(theme_template_id)` or
`(theme_template_id, website_id)` as well (to not have more than one record in `copy_ids`), not just `(key, website_id)`.

**Comment by rdeodoo** [comment]
I don't get it but ok.
You are just confirming what I was thinking, the user has to mess with the backend technical stuff in debug mode.
At that point, he has like billions ways of breaking his database, why are we trying to handle this case, I don't get it..?
And from all the way possible to break a database, this one is not even the easiest one as you have to edit a form view to add a field you are not even supposed to know, duplicate & edit records..

I don't think I can bring anymore useful info here, my opinion was widely explained, I'll let you handle this one as you think it is the best :+1:

**Comment by Yenthe666** [comment]
Agree with @rdeodoo there's hundreds of ways/tricks to break your db if you start doing this kind off things. It's deliberately trying to mess with things 🤷

**Comment by KangOl** [comment]
It doesn't look like the user mess up the attachment directly in the backend
```
chs_42737=> select id, type, name, key, theme_template_id, company_id, original_id from ir_attachment where id in (1025, 794);
  id  |  type  |            name             |         key         | theme_template_id | company_id | original_id
------+--------+-----------------------------+---------------------+-------------------+------------+-------------
  794 | url    | theme_zap.background_img_01 | theme_zap.bg_img_01 |                 1 |          1 |
 1025 | binary | theme_zap.background_img_01 | theme_zap.bg_img_01 |                 1 |          1 |         794
(2 rows)

chs_42737=>
```

So this situation can appear in version 14.0

**Comment by diagnoza** [comment]
imho the actual bug is that we're clearly missing unique constraint(s) on those fields. 
Now we can't fix it in `14.0` and given that this problem occurred several times already, we can expect more cases like this. I agree that it's always possible to break a database in debug mode, but the vast majority of such cases are individual cases that can be handled in a specific script.

Handling this in a `0.0.0` migration script is problematic because it means that such code would be running for versions where this has already been fixed (presumably since `saas~15.1` onward). Given this and that the bug is in the source code (missing constraints), I'd rather have this admitted as possible and handled in the source.

**Comment by rdeodoo** [comment]
Thank you @KangOl it seems (I'll confirm) this was done through the image edition of the website builder and not by messing in the backend, which could only be detected by having access to the database :+1: 
Nice find.

I'll investigate a bit with this new information.

**Comment by Yenthe666** [comment]
What does the "Last Updated by" show then? 🤔

**Comment by rdeodoo** [comment]
Ok I figured things out. Given @KangOl SQL output:
```
  id  |  type  |            name             |         key         | theme_template_id | company_id | original_id
------+--------+-----------------------------+---------------------+-------------------+------------+-------------
  794 | url    | theme_zap.background_img_01 | theme_zap.bg_img_01 |                 1 |          1 |
 1025 | binary | theme_zap.background_img_01 | theme_zap.bg_img_01 |                 1 |          1 |         794
```

It shouldn't be possible for a duplicated attachment to have both a `original_id` and a `theme_template_id`, as the `original_id` being set means that attachment was created through image edition tools in the website builder such as image CROP.
In that case, we copy the original attachment and store its id in `original_id`.
When copying an attachment, the `theme_template_id` is not supposed to be copied.
But it was not always true in 12.0, see https://github.com/odoo/odoo/commit/254d40ffcc21f56a80a1814da7f6355d44e5670b

So indeed, databases coming from 12.0 could have that "broken database state", which is not possible anymore.

In that regard, is there a way to clean those DB in migration scripts (eg remove theme_template_id if original_id is set) rather than in our codebase, or is it too late?

I guess we would need to do the same script in 13.0 14.0 and 15.0 in migration script ? As this should be a one liner, isn't it better to add those in migration rather than this current PR?

**Comment by KangOl** [comment]
>  But it was not always true in 12.0, see [254d40f](https://github.com/odoo/odoo/commit/254d40ffcc21f56a80a1814da7f6355d44e5670b)

From the commit you linked:
> The issue is more visible from 14.0 where applying some modifications to
images via the editor (crop / filter / optimization / ...) will
duplicate the original image before modifying it.

So it also happened in v14.

> In that regard, is there a way to clean those DB in migration scripts (eg remove theme_template_id if original_id is set) rather than in our codebase, or is it too late?

We can, but the bug still happens if a user updates a module, without doing an upgrade.

Even if we fixup the data during upgrade, this can be fixed via this simple patch
```patch
diff --git addons/website/models/ir_module_module.py addons/website/models/ir_module_module.py
index 62cddbf38a36..0f51c667bec6 100644
--- addons/website/models/ir_module_module.py
+++ addons/website/models/ir_module_module.py
@@ -152,7 +152,7 @@ class IrModuleModule(models.Model):
                 # special case for attachment
                 # if module B override attachment from dependence A, we update it
                 if not find and model_name == 'ir.attachment':
-                    find = rec.copy_ids.search([('key', '=', rec.key), ('website_id', '=', website.id)])
+                    find = rec.copy_ids.search([('key', '=', rec.key), ('website_id', '=', website.id), ("original_id", "=", False)])
 
                 if find:
                     imd = self.env['ir.model.data'].search([('model', '=', find._name), ('res_id', '=', find.id)])
```

OTOH, I don't know how to handle the copied menus/pages/views. They however need to be duplicated from the backend, which is less common and can be handled by custom upgrade scripts.

**Comment by rdeodoo** [comment]
> So it also happened in v14.

Correct. Sorry, I was a bit too fast here, it is indeed a matter of "before this date" than "before this version".

> this can be fixed via this simple patch

Great, LGTM :+1: and if the migration clean such databases, we can remove that code in 15.0 as no one has been migrated yet (and probably add the SQL constraints).

> OTOH, I don't know how to handle the copied menus/pages/views. They however need to be duplicated from the backend, which is less common and can be handled by custom upgrade scripts.

From what I know:
- **Menus**: There is **no way to duplicate a menu** other than from the backend, and most importantly **no theme is creating a website_menu**, so a singleton traceback here is impossible (unless a community theme created some website menu and then the user messed his db in backend)
- **Pages**: Page can be duplicated from the page manager `/website/pages` but **no theme is creating a website_page**, so a singleton traceback here is impossible (unless a community theme created some website menu and then the user messed his db in backend)
- **Views**: Can be duplicated through a lot of ways, but not a single way would concern a view created from a theme as the views created by themes are snippet xpath or assets xpath, which obviously no flow lead to a copy of those views. Views can also be duplicated when duplicating a page but see above, not a real issue.

**Comment by KangOl** [comment]
> I am ok with your initial fix if you really want it.

Actually, the current patch of this PR is wrong. We should not update the copies the user made (explicitly or view the editor). So we should ensure `len(find) <= 1`. We should take the record with the earlier creation date
```python
if len(find) > 1:
    find = find.sorted('create_date')[0]
```

However, the search of the XMLID should be strengthened: we can't assume there will always be one and only one xmlid.

**Comment by diagnoza** [comment]
Hey @rdeodoo, @KangOl,

I pushed the proposed changes. Tested on impacted requests.

> Either way, we should add a migration script at least in 15.0 to remove the theme_template_id from attachment when original_id is set[1]. And also add an SQL constraint.
> Let us know if you want us to do it.

I can write a script that does this. When it's done I'll reference this PR and add you as a reviewer @rdeodoo, is that okay?

File: addons/website/models/ir_module_module.py
Comment (rdeodoo) [inline addons/website/models/ir_module_module.py:158]:
```suggestion
```

File: addons/website/models/ir_module_module.py
Comment (rdeodoo) [inline addons/website/models/ir_module_module.py:162]:
```suggestion
                    # In master, an unique constraint over (theme_template_id, website_id)
```

File: addons/website/models/ir_module_module.py
Comment (rdeodoo) [inline addons/website/models/ir_module_module.py:165]: Can't you do that in the search directly so we just have a one line diff?
```python
find = rec.copy_ids.search([('key', '=', rec.key), ('website_id', '=', website.id), ("original_id", "=", False)], order='create_date ASC', limit=1)
```

File: addons/website/models/ir_module_module.py
Comment (rdeodoo) [inline addons/website/models/ir_module_module.py:161]: Let's not talk about that flow, we don't really support those backend tweak flow.
What we are fixing here is just the legit flows duplicating attachment (eg image CROP) that "broke" the DB  **before** the fix copy=False on theme_template_id was applied (even if the backend tweak flow will also be fixed by that, but that's not our goal).

**Review by rdeodoo — CHANGES_REQUESTED** [review]
> I can write a script that does this. When it's done I'll reference this PR and add you as a reviewer @rdeodoo, is that okay?

Sure.
In upgrade (0.0.0 you said?):
- Remove theme_template_id if original_id is set -> Be clear that it only aims to fix the DB that did copy an attachment through flows like CROP image in website builder, prior to the fix that added `copy=false` on theme_template_id field.

In community master:
- Add an SQL constraint on `website_id/theme_template_id` -> Key is not needed?
- Remove the code introduced here as it would not be needed anymore

**Comment by diagnoza** [comment]
> In upgrade (0.0.0 you said?):

Yup, a script that'd be run for every database.

> Add an SQL constraint on website_id/theme_template_id -> Key is not needed?

Well, I gave it some thought and I believe the constraint should also include `key`. The following search:

`find = rec.copy_ids.search([('key', '=', rec.key), ('website_id', '=', website.id), ("original_id", "=", False)], order='create_date ASC', limit=1)`

does not include `theme_template_id`, so, if not for `limit=1`, `find` could still be more than a single record. Now take it with a grain of salt, I feel like I may be be missing here something. But again, I cannot think of such scenario taking place, other than the user messing with the database the way he shouldn't be.

**Comment by rdeodoo** [comment]
> does not include theme_template_id

It does, `rec.copy_ids` is the o2m of the theme_template_id m2o

**Comment by diagnoza** [comment]
Hey @KangOl @rdeodoo may I get a re-review on this? cheers

File: addons/website/models/ir_module_module.py
Comment (rdeodoo) [inline addons/website/models/ir_module_module.py:157]: Now that we figured the issue and know it comes from the fact that before [this filx](https://github.com/odoo/odoo/commit/254d40ffcc21f56a80a1814da7f6355d44e5670b), some records could be copied with their `theme_template_id` (making it also copies) through the website editor, especially with image manipulation like crop. Why don't you just add the real fix which is `("original_id", "=", False)` (like @KangOl suggested back then)?

The order and the limit are useless, you will only get one result now that you exclude the "broken" record(s) by inserting `("original_id", "=", False)`, isn't it?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Could you please:
- squash your commits, the first one is being reverted by the second one
- write a commit message (there is none for now?), [my summary](https://github.com/odoo/odoo/pull/79197#issuecomment-961062591) could be helpful.

And hopefully we don't forget to not merge this in master but revert this commit and introduce the SQL constraint.

**Comment by rdeodoo** [comment]
Oh, you could also mention https://github.com/odoo/upgrade/pull/3110 in your commit message, to point out that on top of this codebase fix, this issue is now getting fixed during the migration.

**Review by rdeodoo — CHANGES_REQUESTED** [review]
LGTM once the commit message got changed...

image

72 max character per lines please :see_no_evil:
Edit: Or 80 at least

**Comment by diagnoza** [comment]
> LGTM once the commit message got changed...
> 
> image
> 
> 72 max character per lines please see_no_evil Edit: Or 80 at least

Should be good now.

Thanks for all the feedback @rdeodoo @KangOl @Yenthe666, this PR showed me that I've still much to learn about Odoo :p

**Review by rdeodoo — APPROVED** [review]
Thank you, AFAIC, this is fine now.

We will get rid of it in master and introduce the SQL Constraint.

I'll let @KangOl have the final look and r+

**Comment by KangOl** [comment]
@robodoo r+

**Comment by rdeodoo** [comment]
Great, I think every migration pending PR regarding website have been handled/merged. :partying_face:

**Pull request closed without merging**
